lightningnetwork/lnd
# Only log insufficient bandwidth warnings if the htlc fails

Previously, a Warning was logged for each channel link to a peer that had insufficient bandwidth (or violated another policy parameter). This seemed excessive and unhelpful because warnings for insufficient balance would be emitted even for htlcs that were successfully routed (see [Issue #5972](https://github.com/lightningnetwork/lnd/issues/5972)).

Also make each error case of htlcswitch.channelLink::canSendHtlc emit a DetailedLinkError rather than a LinkError.

## Change Description
Description of change / link to associated issue.

## Steps to Test

Success:
1. Have multiple channels to a peer, only one with lots of local balance, the others with low local balance
2. Receive forwarding request towards peer for amount that can only be satisfied by one of the channels.
3. Ensure that no "insufficient balance" warning is logged. Debug logs should still appear for each of the low-balance channels.

Failure Case:
1. Have multiple channels to a peer, only one with lots of local balance, the others with low local balance
2. Receive a forwarding request towards peer for amount greater than your largest channel.
3. Ensure that a single "insufficient balance" Warning is logged and it specifies the Channel Point of the channel with large local balance. 

## Pull Request Checklist
### Testing

### Code Style and Documentation

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.
